File: klipper_docs_mcp/services/storage_manager.py
"""Storage Manager for Klipper Docs MCP Server.

Handles file system operations including reading files and listing documentation structure.
"""

import os
from pathlib import Path
from typing import Optional

from ..config import StorageConfig
from ..exceptions import (
    DocumentationNotAvailableError,
    InvalidPathError,
    PathTraversalError,
    ResourceNotFoundError,
)


class StorageManager:
    """Manages file system operations for documentation access."""

    def __init__(self, config: Optional[StorageConfig] = None):
        """Initialize the storage manager.

        Args:
            config: Storage configuration. If None, uses default config.
        """
        self.config = config or StorageConfig()

    @property
    def docs_dir(self) -> Path:
        """Get the documentation directory path."""
        return self.config.docs_dir

    def is_available(self) -> bool:
        """Check if documentation directory exists and is accessible."""
        return self.docs_dir.exists()

    def require_available(self) -> None:
        """Raise an exception if documentation is not available."""
        if not self.is_available():
            raise DocumentationNotAvailableError()

    def validate_path(self, path: str) -> Path:
        """Validate and resolve a path, ensuring it's within docs directory.

        Args:
            path: Relative path to validate.

        Returns:
            Resolved absolute path.

        Raises:
            InvalidPathError: If path is invalid or outside docs directory.
            PathTraversalError: If path traversal attempt is detected.
        """
        target_path = (self.docs_dir / path).resolve()

        try:
            common_path = os.path.commonpath([self.docs_dir, target_path])
            if common_path != str(self.docs_dir):
                raise PathTraversalError(path)
        except ValueError:
            raise InvalidPathError(path, reason="invalid path")

        return target_path

    def read_file(self, path: str, offset: int = 0, limit: Optional[int] = None) -> tuple[str, int]:
        """Read a documentation file.

        Args:
            path: Relative path to the file.
            offset: Character offset to start reading from.
            limit: Maximum number of characters to read.

        Returns:
            Tuple of (content, total_chars).

        Raises:
            DocumentationNotAvailableError: If docs directory doesn't exist.
            ResourceNotFoundError: If the specified file doesn't exist.
            InvalidPathError: If path is invalid.
        """
        self.require_available()
        limit = limit or self.config.max_file_chars

        target_path = self.validate_path(path)

        if not target_path.exists():
            raise ResourceNotFoundError(path)

        with open(target_path, "r", encoding="utf-8", errors="ignore") as f:
            content = f.read()

        total_chars = len(content)
        end = offset + limit
        content_slice = content[offset:end]

        return content_slice, total_chars

    def list_files(self) -> list[Path]:
        """List all documentation files.

        Returns:
            List of relative paths to documentation files.

        Raises:
            DocumentationNotAvailableError: If docs directory doesn't exist.
        """
        self.require_available()

        files = []
        for root, dirs, filenames in os.walk(self.docs_dir):
            for filename in filenames:
                if filename.endswith(self.config.supported_extensions):
                    full_path = Path(root) / filename
                    rel_path = full_path.relative_to(self.docs_dir)
                    files.append(rel_path)

        return files

    def build_tree(self, path: Optional[Path] = None, prefix: str = "", is_last: bool = True) -> list[str]:
        """Recursively build directory tree representation.

        Args:
            path: Path to build tree from. Defaults to docs_dir.
            prefix: Prefix for each line (used for recursion).
            is_last: Whether current entry is the last in its parent.

        Returns:
            List of strings representing the tree structure.
        """
        if path is None:
            path = self.docs_dir

        lines = []
        try:
            entries = sorted(path.iterdir(), key=lambda x: (not x.is_dir(), x.name.lower()))
            entries = [e for e in entries if not e.name.startswith(".")]
        except PermissionError:
            return lines

        for i, entry in enumerate(entries):
            is_last_entry = i == len(entries) - 1
            connector = "└── " if is_last_entry else "├── "

            if entry.is_dir():
                lines.append(f"{prefix}{connector}{entry.name}/")
                extension = "    " if is_last_entry else "│   "
                lines.extend(self.build_tree(entry, prefix + extension, is_last_entry))
            else:
                lines.append(f"{prefix}{connector}{entry.name}")

        return lines

File: src/services/StorageManager.ts
import fs from "node:fs/promises";
import path from "node:path";
import { DOCS_DIR, MAX_FILE_CHARS, SUPPORTED_EXTENSIONS } from "../config.js";
import {
  DocumentationNotAvailableError,
  InvalidPathError,
  PathTraversalError,
  ResourceNotFoundError,
} from "../errors.js";

export class StorageManager {
  private docsDir: string;

  constructor(docsDir: string = DOCS_DIR) {
    this.docsDir = docsDir;
  }

  getDocsDir(): string {
    return this.docsDir;
  }

  async isAvailable(): Promise<boolean> {
    try {
      const stats = await fs.stat(this.docsDir);
      return stats.isDirectory();
    } catch {
      return false;
    }
  }

  async requireAvailable(): Promise<void> {
    if (!(await this.isAvailable())) {
      throw new DocumentationNotAvailableError();
    }
  }

  validatePath(relativePath: string): string {
    const targetPath = path.resolve(this.docsDir, relativePath);

    // Ensure the path starts with docsDir
    if (!targetPath.startsWith(this.docsDir)) {
      throw new PathTraversalError(relativePath);
    }

    return targetPath;
  }

  async readFile(
    relativePath: string,
    offset: number = 0,
    limit: number = MAX_FILE_CHARS
  ): Promise<{ content: string; totalChars: number }> {
    await this.requireAvailable();

    const targetPath = this.validatePath(relativePath);

    try {
      const contentBuffer = await fs.readFile(targetPath, "utf-8");
      // Normalize line endings if needed, but simple read is usually fine
      const content = contentBuffer.toString();
      const totalChars = content.length;
      const end = offset + limit;
      const contentSlice = content.slice(offset, end);

      return { content: contentSlice, totalChars };
    } catch (error: any) {
      if (error.code === "ENOENT") {
        throw new ResourceNotFoundError(relativePath);
      }
      throw new InvalidPathError(relativePath, error.message);
    }
  }

  async listFiles(): Promise<string[]> {
    await this.requireAvailable();
    const files: string[] = [];

    async function traverse(dir: string, baseDir: string) {
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory()) {
          await traverse(fullPath, baseDir);
        } else if (entry.isFile()) {
           const ext = path.extname(entry.name);
           if (SUPPORTED_EXTENSIONS.includes(ext)) {
               files.push(path.relative(baseDir, fullPath));
           }
        }
      }
    }

    await traverse(this.docsDir, this.docsDir);
    return files;
  }

  async buildTree(currentPath: string = this.docsDir, prefix: string = ""): Promise<string[]> {
    if (!(await this.isAvailable())) {
       return [];
    }

    const lines: string[] = [];
    let entries;
    
    try {
        entries = await fs.readdir(currentPath, { withFileTypes: true });
    } catch (e) {
        return lines;
    }

    // Sort: directories first, then files (case insensitive)
    entries.sort((a, b) => {
        if (a.isDirectory() && !b.isDirectory()) return -1;
        if (!a.isDirectory() && b.isDirectory()) return 1;
        return a.name.localeCompare(b.name, undefined, { sensitivity: 'base' });
    });
    
    // Filter hidden files
    entries = entries.filter(e => !e.name.startsWith('.'));

    for (let i = 0; i < entries.length; i++) {
        const entry = entries[i];
        const isLastEntry = i === entries.length - 1;
        const connector = isLastEntry ? "└── " : "├── ";
        
        if (entry.isDirectory()) {
            lines.push(`${prefix}${connector}${entry.name}/`);
            const extension = isLastEntry ? "    " : "│   ";
            const subLines = await this.buildTree(path.join(currentPath, entry.name), prefix + extension);
            lines.push(...subLines);
        } else {
             lines.push(`${prefix}${connector}${entry.name}`);
        }
    }
    return lines;
  }
}
